tiledb/sm/cpp_api/arrow_io_impl.h

```diff
@@ -0,0 +1,832 @@
+  }
+
+  /*
+   * CPPArrowSchema destructor
+   *
+   * This destructor is invoked via the ArrowSchema.release
+   * callback. Owned member data is released via default destructors.
+   */
+  ~CPPArrowSchema() {
+    if (schema_ != nullptr)
+      std::free(schema_);
+  };
+
+  /*
+   * Exports the ArrowSchema to a pre-allocated target struct
+   *
+   * This function frees the array_.
+   * The lifetime of all other member variables is controlled
+   * by the ArrowSchema.release callback, which frees the
+   * CPPArrowSchema structure (via ArrowSchema.private_data).
+   */
+  void export_ptr(ArrowSchema* schema_out) {
+    memcpy(schema_out, schema_, sizeof(ArrowSchema));
+    free(schema_);
+    schema_ = nullptr;
+  }
+
+  ArrowSchema* mutable_ptr() {
+    assert(schema_ != nullptr);
+    return schema_;
+  }
+
+  ArrowSchema* ptr() const {
+    assert(schema_ != nullptr);
+    return schema_;
+  }
+
+ private:
+  ArrowSchema* schema_;
+  std::string format_;
+  std::string name_;
+  std::string metadata_;
+  int64_t flags_;
+  int64_t n_children_;
+  std::vector<ArrowSchema*> children_;
+  std::shared_ptr<CPPArrowSchema> dictionary_;
+};
+
+struct CPPArrowArray {
+  /*
+   * Initialize a CPPArrowSchema object
+   *
+   * The lifetime of this object is controlled by the
+   * release callback set in the ArrowSchema.
+   *
+   * Note that an ArrowSchema is *movable*, provided
+   * the release callback of the source is set to null.
+   */
+  CPPArrowArray(
+      int64_t elem_num,
+      int64_t null_num,
+      int64_t offset,
+      std::vector<std::shared_ptr<CPPArrowArray>> children,
+      std::vector<void*> buffers) {
+    array_ = static_cast<ArrowArray*>(std::malloc(sizeof(ArrowArray)));
+    if (array_ == nullptr)
+      throw tiledb::TileDBError("Failed to allocate ArrowArray");
+
+    // Data description
+    array_->length = elem_num;
+    array_->null_count = null_num;
+    array_->offset = offset;
+    array_->n_buffers = (int64_t)buffers.size();
+    array_->n_children = (int64_t)children.size();
+    array_->buffers = nullptr;
+    array_->children = nullptr;
+    array_->dictionary = nullptr;
+    // Bookkeeping
+    array_->release = ([](ArrowArray* array_p) {
+      assert(array_p->release != nullptr);
+
+      // Release children
+      for (int64_t i = 0; i < array_p->n_children; i++) {
+        ArrowArray* child_array = array_p->children[i];
+        child_array->release(child_array);
+        assert(child_array->release == nullptr);
+      }
+
+      // Release dictionary
+      struct ArrowArray* dict = array_p->dictionary;
+      if (dict != NULL && dict->release != NULL) {
+        dict->release(dict);
+        assert(dict->release == NULL);
+      }
+
+      // mark the ArrowArray struct as released
+      array_p->release = nullptr;
+
+      delete (CPPArrowArray*)(array_p->private_data);
+    });
+    array_->private_data = this;
+
+    buffers_ = buffers;
+    array_->buffers = (const void**)buffers_.data();
```

Comment:
switch to `static_cast` instead of c style casting

Reply:
had to use `const_cast` here